src/lua/alloc.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2024, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "lua/utils.h"
+
+#include <lauxlib.h> /* luaL_error */
+
+#define LUA_MEMORY_DEFAULT (2ULL * 1024 * 1024 * 1024)
+
+/**
+ * The default allocator function which is wrapped into a new one
+ * with the Lua memory limit checker.
+ */
+static lua_Alloc orig_alloc;
+/**
+ * The maximum allowed LuaJIT memory limit.
+ */
+static size_t memory = LUA_MEMORY_DEFAULT;
+/**
+ * Currently allocated memory amount by LuaJIT.
+ */
+static size_t allocated;
+
+/**
+ * Lua custom memory allocation function. It extends the original
+ * one with a memory counter and a limit check.
+ */
+static void *
+luaT_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+	size_t new_allocated = allocated + nsize - osize;
+	if (new_allocated > memory) {
+		/* Returning NULL results with "not enough memory" error. */
+		return NULL;
+	}
+
+	void *result = orig_alloc(ud, ptr, osize, nsize);
+
+	if (result != NULL || nsize == 0)
+		allocated = new_allocated;
+
+	/*
+	 * Result may be NULL, in this case "not enough memory"
+	 * is raised.
+	 */
+	return result;
+}
+
+int
+luaT_setallocmemory(size_t amount)
+{
+	if (amount < allocated) {
+		diag_set(LuajitError, "Cannot limit the Lua memory with values "
+			 "less than the currently allocated amount");
+		return -1;
+	}
+	memory = amount;
+	return 0;
+}
+
+void
+luaT_initalloc(struct lua_State *L)
+{
+	assert(L != NULL);
+
+	allocated = luaL_getgctotal(L);
+
+	void *orig_ud;
+	orig_alloc = lua_getallocf(L, &orig_ud);
+
+	assert(orig_alloc != NULL);
+
+	lua_setallocf(L, luaT_alloc, orig_ud);
+}
+
+/**
+ * alloc.getmemory() -- get the allocator memory limit.
+ */
+static int
+lbox_alloc_getmemory(struct lua_State *L)
+{
+	lua_pushinteger(L, memory);
+	return 1;
+}
+
+/**
+ * alloc.setmemory() -- set the allocator memory limit.
+ * Returns old memory limit.
+ */
+static int
+lbox_alloc_setmemory(struct lua_State *L)
+{
+	if (lua_gettop(L) < 1) {
+		diag_set(IllegalParams, "Usage: alloc.setmemory(amount)");
+		return luaT_error(L);
+	}
+
+	ssize_t amount = luaL_checkinteger(L, 1);
+
+	if (amount < 0) {
+		diag_set(IllegalParams, "Invalid memory amount: "
+			 "the value must be >= 0");
+		return luaT_error(L);
+	}
+
+	size_t old_memory = memory;
+
+	if (luaT_setallocmemory(amount) < 0)
+		return luaT_error(L);
+
+	lua_pushinteger(L, old_memory);
+	return 1;
+}
+
+/**
+ * alloc.allocated() -- get the amount of allocated memory.
+ */
+static int
+lbox_alloc_allocated(struct lua_State *L)
+{
+	lua_pushinteger(L, allocated);
+	return 1;
+}
+
+/**
+ * alloc.allocated() -- get the amount of free memory.
+ */
```

Comment:
Typo: alloc.allocated() -> alloc.free().

Reply:
Now the method is called `alloc.unused()`. Applied.